plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,620 @@
+ */
+static gboolean
+fu_dell_dock_usb4_mbox_data_read (FuDevice *device, guchar *data, guint8 length,
+				  GError **error)
+{
+	guchar *ptr = data;
+	gint	i;
+
+	if (length > 64 || length % 4) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "invalid firmware data read length %u",
+			     length);
+		return FALSE;
+	}
+	/* read 4 bytes per iteration */
+	for (i = 0; i < length / 4; i++) {
+		if (!fu_dell_dock_usb4_hub_get_mmio (device, i, ptr, error))
+			return FALSE; /* FIXME also print a error message */
+		ptr += 4;
+	}
+	return TRUE;
+}
+
+/*
+ * The mailbox can hold 64 bytes in 16 doubleword data registers.
+ * A NVM write operation writes data from these registers to NVM
+ * at the set offset
+ */
+static gboolean
+fu_dell_dock_usb4_mbox_data_write (FuDevice *device, const guchar *data,
+				   guint8 length, GError **error)
+{
+	guchar *ptr = (guchar *) data;
+
+	if (length > 64 || length % 4) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "invalid firmware data write length %u\n",
+			     length);
+		return FALSE;
+	}
+	/* writes 4 bytes per iteration */
+	for (gint i = 0; i < length / 4; i++) {
+		if (!fu_dell_dock_usb4_hub_set_mmio (device, i, ptr, error))
+			return FALSE;
+		ptr += 4;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_hub_operation (FuDevice *device, guint16 opcode,
+				 guchar *metadata, guint8 *status,
+				 GError **error)
+{
+	struct mbox_regx *regx;
+	guchar buf[4];
+	gint   i;
+
+	regx = (struct mbox_regx *)buf;
+	regx->opcode = GUINT16_TO_LE(opcode);
+	regx->status = MBOX_OPVALID;
+
+	/* Write metadata register for operations that use it */
+	switch (opcode) {
+		case OP_NVM_WRITE:
+		case OP_NVM_AUTH_WRITE:
+			break;
+		case OP_NVM_READ:
+		case OP_NVM_SET_OFFSET:
+		case OP_DROM_READ:
+			if (!metadata) {
+				g_set_error (error,
+					     G_IO_ERROR,
+					     G_IO_ERROR_INVALID_DATA,
+					     "hub opcode 0x%x requires metadata",
+					     opcode);
+				return FALSE;
+			}
+			if (!fu_dell_dock_usb4_hub_set_mmio (device,
+							     MBOX_REG_METADATA,
+							     metadata, error)) {
+				g_prefix_error (error,
+						"failed to write metadata %s:",
+						metadata);
+				return FALSE;
+			}
+			break;
+		default:
+			g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+				     "invalid hub opcode: 0x%x", opcode);
+			return FALSE;
+	}
+
+	/* write the operation and poll completion or error */
+	if (!fu_dell_dock_usb4_hub_set_mmio (device, MBOX_REG, buf, error))
+		return FALSE;
+
+	for (i = 0; i < 100; i++) {
+		if (!fu_dell_dock_usb4_hub_get_mmio (device, MBOX_REG, buf,
+						     error))
+			return FALSE;
+
+		/* Error status bit set, operation failed */
+		if (regx->status & MBOX_ERROR) {
+			g_prefix_error (error,
+					"hub failed to execute operation 0x%x, "
+					"status 0x%x, opcode 0x%x:",
+					opcode, regx->status, regx->opcode);
+			return FALSE; /* FIXME error code */
+		}
```

Comment:
I think this needs to go away, it should be handled by the false return of `fu_dell_dock_usb4_hub_get_mmio`

Reply:
I thought fu_dell_dock_usb4_hub_get_mmio determines if the usb hub control is set successfully, the hub thereafter uses this field to indicate the status of an hub operation after completion. If the error status bit is set it means the opcode operation failed internally for some reason.

Reply:
Handle that all in function. The function failed for some reason read op code in it. Don't have it returned mixed values. Its too confusing this way.